### ISSUE TITLE
Deviation in numerical computations between 32 and 64 bit build (Windows, possibly the same on other OS)

### ISSUE THREAD

**Issue by MSoegtropIMC** [coq/coq#11321]
Dear Coq Team,

on 8.11 master I get deviations in numerical computations between 32 bit and 64 bit Windows builds with Interval. Since Interval is entirely written in Coq (no plugin) this might be an error in numerical computations in the VM and it is quite possible that @silene can construct a proof of False out of this.

The Code is
```
Require Import Reals Coquelicot.Coquelicot.
Require Import Interval.Tactic.

Goal True.
interval_intro (Rabs (RInt (fun x => exp x) 0 3 - (exp(1)*exp(1)*exp(1) - 1))) with (i_integral_depth 0, i_integral_deg 12).
```

On a 64 bit Windows build I get the correct answer:
```
H : 0 <= Rabs (RInt (fun x : R => exp x) 0 3 - (exp 1 * exp 1 * exp 1 - 1)) <= 26 / 33554432
```

On 32 bit Windows I get the wrong answer:
```
H : 6607721 / 33554432 <= Rabs (RInt (fun x : R => exp x) 0 3 - (exp 1 * exp 1 * exp 1 - 1)) <=
    31676929 / 33554432
```

I set this to priority blocker because I think one should at least resarch if a proof of False can be made out of this before the next release or give a warning that 32 bit builds might be unstable.

**Comment by maximedenes** [comment]
Ouch. Is `interval_intro` an ML or an Ltac tactic? It would be nice to see exactly where its behavior starts to diverge between the two architectures.

**Comment by MSoegtropIMC** [comment]
There are no .ml files in the interval source tree, so it is for sure all Ltac and Gallina (mostly Gallina with just a thin Ltac wrapper layer afaik). The dependencies are:

-  mathcomp
-  coquelicot
-  bignums
-  flocq

Afaik for bignums only the syntax is an ML plugin:

```
./BigN/gen/NMake_gen.ml
./plugin/bignums_syntax.ml
```

Mathcomp has two ML files, but afaik they have nothing to do with numeric computation:
```
./etc/utils/dependtodot.ml
./etc/utils/hierarchy.ml
```

Flocq and Coquelicot have no ML files. This makes me believe that it is an issue with the numerics built into VM compute.

I will check if the strip down proofs of False @silene created for the division issue still fail for 32 bit. I am quite sure I tested 32 and 64 bit back then, but maybe I did a mistake.

Anyway I think it is due time to add CoqInterval to CI - it is fairly quick and seems to be a rather strong test for the numerical part of VMcompute (this failure was part of the CoqInterval regression test suite). I can contribute a few additonal really tough tests (which require more than 32 bit final result precision) - this test only requires 20 bit. My bad, I shoudl have done this a long time back.

**Comment by MSoegtropIMC** [comment]
P.S.: I tested the proof of False found by @silene for #10551 to see if this was only fixed for 64 bit:
```
From Bignums Require Import BigN.

Lemma foo (v := BigN.sqrt (BigN.pow 2 151)) : False.
Proof.
generalize (eq_refl v) (BigN.eqb_refl v).
unfold v at 1 3.
clearbody v.
intros H1.
cbv in H1.
elim H1.
vm_compute.
discriminate.
Qed.
```
This proof does not work any more, neither in 32 bit nor in 64 bit Coq.

So this is a different issue.

**Comment by maximedenes** [comment]
You can try `Set Ltac Debug. Set Ltac Batch Debug.` to see where traces diverge in 64 vs 32 bits.

**Comment by silene** [comment]
> I think one should at least resarch if a proof of False can be made out of this

You just proved `6607721 / 33554432 <= 0`. So, unless `Qed` rejects the proof produced by `interval_intro`, you have a trivial proof of `False`. And since the proof produced by `interval_intro` was already checked by the kernel at `abstract` time, there is almost nothing left that can fail at `Qed` time.

> Is `interval_intro` an ML or an Ltac tactic?

Ltac. But more importantly, it is plain computational reflection. So, the whole thing is just a proof of `true = true` (or rather, in the present case, `false = true`).

**Comment by silene** [comment]
You could try passing `i_native` to `interval_intro`. Just so we know whether both `vm_compute` and `native_compute` are broken, or just `vm_compute`.

**Comment by MSoegtropIMC** [comment]
> You could try passing `i_native` to `interval_intro`. Just so we know whether both `vm_compute` and `native_compute` are broken, or just `vm_compute`.

This is on Windows where native compute does not work. I will see if I can make a 32 bit Linux build. 

Btw.: to test Interval on a recent Coq (8.11 or 8.12) you need the patch from #11318 to adjust to the incompatibilities in the Reals library - I was going to send you an email about this.

**Comment by MSoegtropIMC** [comment]
> 6607721 / 33554432 <= 0

Ah yes - I didn't really look at the integrand.

**Comment by silene** [comment]
How confident are we that `Uint63_31.mulc` is correct?

```
# mulc (Int64.of_int 0x1_0000_0000) (Int64.of_int 0x1_0000_0000);;
- : int64 * int64 = (2L, 0L)
# mulc (Int64.of_int 0x8000_0000) (Int64.of_int 0x1_0000_0000);;
- : int64 * int64 = (2L, 0L)
```

**Comment by ppedrot** [comment]
Great, I don't know any better Christmas present than a soundness bug discovered and fixed *before* a release.

-- Truly, your RM.

**Comment by MSoegtropIMC** [comment]
I reopened this since #11323 needs to be back merged into 8.11.

**Comment by Zimmi48** [comment]
@MSoegtropIMC Please don't. That's the RM job and he has all he needs to not forget (namely the backporting column: https://github.com/coq/coq/projects/35#column-7305837)
